File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> map = new HashMap<>();
        int left = 0;
        int right = 0;
        int answer = 0;
        
        while (right < s.length()) {
            char r = s.charAt(right);
            map.put(r, map.getOrDefault(r, 0) + 1);
            while (map.get(r) > 1) {
                char l = s.charAt(left);
                map.put(l, map.get(l) - 1);
                left++;
            }
            answer = Math.max(answer, right - left + 1);
            right++;
        }
        
        return answer;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.kt
class Solution {
    fun lengthOfLongestSubstring(s: String): Int {
        val map = mutableMapOf<Char, Int>()

        var left = 0
        var right = 0
        var answer = 0

        while (right < s.length) {
            val r = s[right]
            map[r] = map.getOrDefault(r, 0) + 1
            while (map.get(r)!! > 1) {
                val l = s[left]
                map[l] = map.get(l)!! - 1
                left++
            }

            answer = Math.max(answer, right - left + 1)
            right++
        }

        return answer
    }
}